{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c5ecad90",
   "metadata": {},
   "source": [
    "# Process Coverage Notebook (12/9/2021)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ca6865",
   "metadata": {},
   "source": [
    "This notebook demonstrates how to load and process .cov/.cov.gz single-cell methylome files generated by [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) into binarized methylation matrices that are used as input to the `run_scAge` function. Please refer to the README file on the GitHub page for information on alternative processing methods for single-cell methylation data. <br><br> Single-cell methylation profiles from the [(Gravina et al, 2016)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1011-3) study are provided as both raw .cov.gz files and processed .tsv.gz files in `sc_data_raw` and `sc_data_processed`, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0a9498",
   "metadata": {},
   "source": [
    "## Import required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28b639ff",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T18:45:13.940903Z",
     "start_time": "2021-06-13T18:45:13.353160Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "sys.path.append('..')\n",
    "import scAge\n",
    "import subprocess\n",
    "import multiprocessing\n",
    "num_total_cores = multiprocessing.cpu_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a3dc93",
   "metadata": {},
   "source": [
    "## Check coverage files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2450fb38",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T18:45:16.765562Z",
     "start_time": "2021-06-13T18:45:16.757667Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coverage file input directory: '../sc_data_raw/'\n",
      "\tRaw .cov file #1 --> 'SRR3136624_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #2 --> 'SRR3136625_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #3 --> 'SRR3136626_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #4 --> 'SRR3136627_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #5 --> 'SRR3136628_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #6 --> 'SRR3136629_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #7 --> 'SRR3136630_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #8 --> 'SRR3136631_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #9 --> 'SRR3136634_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #10 --> 'SRR3136635_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #11 --> 'SRR3136646_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #12 --> 'SRR3136647_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #13 --> 'SRR3136651_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #14 --> 'SRR3136652_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #15 --> 'SRR3136653_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #16 --> 'SRR3136654_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #17 --> 'SRR3136655_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #18 --> 'SRR3136656_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #19 --> 'SRR3136657_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #20 --> 'SRR3136658_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #21 --> 'SRR3136659_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #22 --> 'SRR3136660_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #23 --> 'SRR3136661_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #24 --> 'SRR3136662_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #25 --> 'SRR3136663_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #26 --> 'SRR3136664_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\n",
      "Processed file output directory: '../sc_data_processed/'\n"
     ]
    }
   ],
   "source": [
    "# designate input directory of .cov files\n",
    "input_coverage_directory = \"../sc_data_raw/\"\n",
    "\n",
    "# get .cov files\n",
    "input_coverage_files = sorted(os.listdir(input_coverage_directory))\n",
    "\n",
    "# remove \".ipynb_checkpoints\" file if present\n",
    "if \".ipynb_checkpoints\" in input_coverage_files:\n",
    "    input_coverage_files.remove(\".ipynb_checkpoints\")\n",
    "    os.rmdir(input_coverage_directory + \".ipynb_checkpoints\")\n",
    "print(\"Coverage file input directory: '%s'\" % input_coverage_directory)\n",
    "\n",
    "# cycle through .cov files\n",
    "for count, file in enumerate(input_coverage_files):\n",
    "    print(\"\\tRaw .cov file #%s --> '%s'\" % (count + 1, file))\n",
    "    \n",
    "# denote output path for processed .tsv files\n",
    "output_path = \"../sc_data_processed/\"\n",
    "print(\"\\nProcessed file output directory: '%s'\" % output_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a17fc294",
   "metadata": {},
   "source": [
    "## Process coverage files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "054761bd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T18:47:32.483200Z",
     "start_time": "2021-06-13T18:46:33.841339Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "process_coverage function starting!\n",
      "\n",
      "----------------------------------------------------------\n",
      "Loading .cov files from '../sc_data_raw/'\n",
      "Number of Bismark .cov files = 26\n",
      "First .cov file name: 'SRR3136624_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "----------------------------------------------------------\n",
      "\n",
      "----------------------------------------------------------\n",
      "Starting parallel loading and processing of .cov files...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "139fd9f3c6be4f9dbb9e6e6775649ddf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Single-cell loading progress :   0%|          | 0/26 [00:00<?, ? cell methylomes/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parallel loading complete!\n",
      "Processed binary methylation matrices written to '../sc_data_processed/'\n",
      "----------------------------------------------------------\n",
      "\n",
      "Time elapsed to process coverage files = 30.942 seconds\n",
      "\n",
      "process_coverage run complete!\n",
      "Compressing output .tsv files\n",
      "Binary methylation matrices compressed!\n"
     ]
    }
   ],
   "source": [
    "# run process_coverage\n",
    "scAge.process_coverage(cov_directory = input_coverage_directory,\n",
    "                       n_cores = num_total_cores,\n",
    "                       max_met = 100,\n",
    "                       split = \"_\",\n",
    "                       chunksize = 1,\n",
    "                       binarization = \"round\",\n",
    "                       output_path = output_path)\n",
    "\n",
    "# gzip output files\n",
    "print(\"\\nCompressing output .tsv files\")\n",
    "rm_previous_gzip = subprocess.run(\"rm %s*.gz\" % output_path, shell = True)\n",
    "gzip_out = subprocess.run(\"gzip -v %s*\" % output_path, shell = True)\n",
    "print(\"Binary methylation matrices compressed!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
