{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c5ecad90",
   "metadata": {},
   "source": [
    "# Process Coverage Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ca6865",
   "metadata": {},
   "source": [
    "This notebook demonstrates how to load and process .cov single-cell methylome files generated by [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) into binarized methylation matrices that are required as input to the run_scAge function. Refer to the README file on the GitHub page for information on alternative processing methods for single-cell methylation data. <br><br> Three single-cell methylation profiles from the [(Gravina et al, 2016)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1011-3) study are provided as both raw .cov files and processed .tsv files in `sc_data_raw` and `sc_data_processed`, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0a9498",
   "metadata": {},
   "source": [
    "## Import required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28b639ff",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T05:28:22.707679Z",
     "start_time": "2021-06-13T05:28:22.138709Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import scAge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a3dc93",
   "metadata": {},
   "source": [
    "## Check coverage files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2450fb38",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T05:35:47.233217Z",
     "start_time": "2021-06-13T05:35:47.226463Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coverage file input directory: './sc_data_raw/'\n",
      "\tRaw .cov file #1 --> 'SRR3136627_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #2 --> 'SRR3136628_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\tRaw .cov file #3 --> 'SRR3136659_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "\n",
      "Processed file output directory: './sc_data_processed/'\n"
     ]
    }
   ],
   "source": [
    "# designate input directory of .cov files\n",
    "input_coverage_directory = \"./sc_data_raw/\"\n",
    "\n",
    "# get .cov files\n",
    "input_coverage_files = sorted(os.listdir(input_coverage_directory))\n",
    "print(\"Coverage file input directory: '%s'\" % input_coverage_directory)\n",
    "\n",
    "# cycle through .cov files\n",
    "for count, file in enumerate(input_coverage_files):\n",
    "    print(\"\\tRaw .cov file #%s --> '%s'\" % (count + 1, file))\n",
    "    \n",
    "# denote output path for processed .tsv files\n",
    "output_path = \"./sc_data_processed/\"\n",
    "print(\"\\nProcessed file output directory: '%s'\" % output_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a17fc294",
   "metadata": {},
   "source": [
    "## Process coverage files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "054761bd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-13T05:34:06.954695Z",
     "start_time": "2021-06-13T05:33:53.509300Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "process_coverage function starting!\n",
      "\n",
      "----------------------------------------------------------\n",
      "Loading .cov files from './sc_data_raw/'\n",
      "Number of Bismark .cov files = 3\n",
      "First .cov file name: 'SRR3136627_bismark_bt2.deduplicated.bismark.cov.gz'\n",
      "----------------------------------------------------------\n",
      "\n",
      "----------------------------------------------------------\n",
      "Starting parallel loading and processing of .cov files...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "68908d8b78bb4f1e91466cddf9c788d1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Single-cell loading progress :   0%|          | 0/3 [00:00<?, ? cell methylomes/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parallel loading complete!\n",
      "Processed binary methylation matrices written to './sc_data_processed/'\n",
      "----------------------------------------------------------\n",
      "\n",
      "Time elapsed to process coverage files = 13.429 seconds\n",
      "\n",
      "process_coverage run complete!\n"
     ]
    }
   ],
   "source": [
    "# run process_coverage\n",
    "scAge.process_coverage(cov_directory = input_coverage_directory,\n",
    "                       n_cores = 3,\n",
    "                       max_met = 100,\n",
    "                       split = \"_\",\n",
    "                       chunksize = 1,\n",
    "                       binarization = \"round\",\n",
    "                       output_path = output_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
